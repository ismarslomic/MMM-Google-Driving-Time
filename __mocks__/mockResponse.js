const generateResponse = () => ({
  status: 200,
  statusText: 'OK',
  headers: {},
  config: {},
  request: {},
  data: {
    'destination_addresses': [
      'GÃ¸teborggata 32, 0566 Oslo, Norge'
    ],
    'origin_addresses': [
      'Hemsilbreidda 16, 3560 Hemsedal, Norge'
    ],
    'rows': [
      {
        'elements': [
          {
            'distance': {
              'text': '203 km',
              'value': 203466
            },
            'duration': {
              'text': '3 timer 0 min',
              'value': 10782
            },
            'duration_in_traffic': {
              'text': '2 timer 48 min',
              'value': 10064
            },
            'status': 'OK'
          }
        ]
      }
    ],
    'status': 'OK'
  }
})

module.exports = { generateResponse }

