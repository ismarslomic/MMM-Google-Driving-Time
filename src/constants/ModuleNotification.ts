export enum ModuleNotification {
  DRIVING_TIME_REQUEST = 'DRIVING_TIME_REQUEST',
  DRIVING_TIME_RESPONSE = 'DRIVING_TIME_RESPONSE'
}
