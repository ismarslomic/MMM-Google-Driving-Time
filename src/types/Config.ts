export type Config = {
  lastUpdateFormat?: string
  updateIntervalInSeconds?: number
  origin: string
  destination: string
  language: string
  trafficMode: 'best_guess' | 'pessimistic' | 'optimistic'
  apiKey: string
}
