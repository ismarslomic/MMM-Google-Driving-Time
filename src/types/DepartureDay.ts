export enum DepartureDay {
  TODAY = 'TODAY',
  TOMORROW = 'TOMORROW'
}
