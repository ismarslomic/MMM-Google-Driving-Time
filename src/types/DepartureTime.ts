export type DepartureTime = {
  hours: number
  minutes: number
}
