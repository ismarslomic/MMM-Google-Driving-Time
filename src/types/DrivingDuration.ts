/**
 * The length of time it takes to drive the route, expressed in seconds (inSeconds) and as text (description).
 * The textual representation is localized according to the query's language parameter.
 */
export type DrivingDuration = {
  /**
   * The length of time in seconds
   */
  inSeconds: number,
  /**
   * The length of time as localized text
   */
  description: string
}
