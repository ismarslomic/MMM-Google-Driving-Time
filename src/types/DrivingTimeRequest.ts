import { DepartureTime } from './DepartureTime'
import { DepartureDay } from './DepartureDay'

export type DrivingTimeRequest = {
  origin: string
  destination: string
  language: string
  trafficModel: 'best_guess' | 'pessimistic' | 'optimistic'
  departureDate: DepartureDay
  departureTimes: DepartureTime[]
  apiKey: string
}
