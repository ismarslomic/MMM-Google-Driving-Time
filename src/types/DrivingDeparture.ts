/**
 * The length of time it takes to drive the route on specific departure time.
 */
import { DrivingDuration } from './DrivingDuration'

export type DrivingDeparture = {
  /**
   * Departure time (date and time) in future
   */
  departureTime: Date
  /**
   * The length of time it takes to drive this route, based on current and historical traffic conditions.
   */
  durationInTraffic: DrivingDuration
  /**
   * Difference (in seconds) between the length of time it takes to travel this route normally and the
   * length of time it takes to travel this route based on current and historical traffic conditions
   *
   * Negative duration means shorter time to travel based on current and historical traffic conditions, positive means
   * longer.
   */
  durationDiff: {
    /**
     * Difference in time duration expressed in seconds, negative means shorter time to travel then normally
     */
    inSeconds: number
    /**
     * Difference in time duration expressed as text, localized according to the query's language parameter.
     */
    description: string
  }
}
