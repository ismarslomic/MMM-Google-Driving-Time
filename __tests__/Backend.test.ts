import { generateResponse } from '../__mocks__/mockResponse'
import { Config } from '../src/types/Config'
import { Client } from '@googlemaps/google-maps-services-js'
import { ModuleNotification } from '../src/constants/ModuleNotification'
import { DepartureDay } from '../src/types/DepartureDay'

describe('Backend', () => {
  let helper
  const mockedDistancematrix = jest.fn()
  let mockedSendSocketNotification
  let config: Config

  beforeEach(() => {
    helper = require('../src/backend/Backend')
    helper.setName('MMM-Google-Driving-Time')

    // Mock the Google Distance Matrix API and return mocked response
    Client.prototype.distancematrix = mockedDistancematrix
    mockedDistancematrix.mockReturnValue(Promise.resolve(generateResponse()))

    // Mock the MMM sendSocketNotification function which returns data back to the frontend
    mockedSendSocketNotification = helper.sendSocketNotification.mock

    config = {
      origin: 'place_id:ChIJt8uR5A3pP0YRFsY8zU7Bg5g',
      destination: 'place_id:ChIJMYz3QEJuQUYRLRlABAEinJA',
      language: 'no',
      trafficModel: 'best_guess',
      apiKey: 'abcdefg',
      departureDate: DepartureDay.TODAY,
      departureTimes: [{ hours: 10, minutes: 0 }]
    }

    jest.useFakeTimers().setSystemTime(new Date('2022-05-01T09:00:00'))
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  test('GIVEN starting the module THEN log starting module to console', () => {
    helper.start()

    expect(console.log).toHaveBeenCalledWith('Starting module helper: MMM-Google-Driving-Time')
  })

  test('GIVEN requesting driving time WHEN response is successful THEN response is sent via socket', async () => {
    helper.socketNotificationReceived(ModuleNotification.DRIVING_TIME_REQUEST, config)
    jest.useRealTimers()
    await global.waitForAsync()

    expect(mockedSendSocketNotification.calls[0][0]).toBe(ModuleNotification.DRIVING_TIME_RESPONSE)
    expect(mockedSendSocketNotification.calls[0][1]).toMatchSnapshot({
      lastUpdate: expect.any(Number),
      drivingTime: {
        drivingDepartures: [
          {
            departureTime: expect.any(Date)
          }
        ]
      }
    })
  })
})
