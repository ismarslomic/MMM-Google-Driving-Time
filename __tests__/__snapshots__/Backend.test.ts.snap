// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backend GIVEN requesting driving time WHEN response is successful THEN response is sent via socket 1`] = `
Object {
  "drivingTime": Object {
    "destination": "GÃ¸teborggata 32, 0566 Oslo, Norge",
    "distance": Object {
      "description": "203 km",
      "inMeter": 203466,
    },
    "drivingDepartures": Array [
      Object {
        "departureTime": Any<Date>,
        "durationDiff": Object {
          "description": "-12 min",
          "inSeconds": -718,
        },
        "durationInTraffic": Object {
          "description": "2 timer 48 min",
          "inSeconds": 10064,
        },
      },
    ],
    "duration": Object {
      "description": "3 timer 0 min",
      "inSeconds": 10782,
    },
    "origin": "Hemsilbreidda 16, 3560 Hemsedal, Norge",
  },
  "lastUpdate": Any<Number>,
}
`;
